src/Qir/Common/Include/qsharp__foundation_internal.hpp

```diff
@@ -6,18 +6,20 @@
 // To be included by the QIS implementation and QIS tests only.
 // Not to be included by parties outside QIS.
 
+#include "CoreTypes.hpp"
+
 // For test purposes only:
 namespace Quantum // Replace with `namespace Quantum::Qis::Internal` after migration to C++17.
 {
 namespace Qis
 {
     namespace Internal
     {
-        extern char const excStrDrawRandomVal[];
+        QIR_SHARED_API extern char const excStrDrawRandomVal[];
 
-        void RandomizeSeed(bool randomize);
-        int64_t GetLastGeneratedRandomI64();
-        double GetLastGeneratedRandomDouble();
+        QIR_SHARED_API void RandomizeSeed(bool randomize);
+        QIR_SHARED_API int64_t GetLastGeneratedRandomI64();
+        QIR_SHARED_API double GetLastGeneratedRandomDouble();
     } // namespace Internal
 } // namespace Qis
 } // namespace Quantum
```

Comment:
(Up to you)
FYI: Originally the `Qiantum::Qis::Internal`s APIs were supposed to be used by the tests only (and not exposed externally to the users).
With recent mass refactoring PRs I'm not sure if that is possible to retain. ;-)

Reply:
Yeah, I couldn't have the test be separate from the runtime without making these `dllexport`. The good news is that by keeping the header out of the src/Qir/Runtime/public folder it's still not something we would expect any code external to our repo to utilize, but someone could copy the header and then access these functions if they were sufficiently motivated. I think we have two choices about how to proceed but I didn't want to wrap that into this change at the same time. Here are my ideas, and I think I'd defer to you on which would be preferable:

1.  Refactor the tests to be true unit tests that don't depend on any Q# compilation, such that they would only all into the random number generation code directly and either not use QIR or have simple hard-coded QIR wrappers. Then we could move it to the src/Qir/Runtime/unittests folder and put this header back into that part of the tree, removing the `dllexport` in the process.
2. Keep the tests where they are but refactor them such that they don't depend on any internal functionality. This would entail duplicating the error string into the test and then updating the test strategy to generate a spread of random numbers and confirm they are not the same, which would have a non-zero chance of failing if the two equal numbers are coincidentally generated. Given how much grief random tests have caused for us in the past, I think there is value in instead considering keeping `RandomizeSeed` as a part of the public API so that this and other tests can request a non-random seed and test with a well known sequence of results.
3. Some combination of both strategies, where the internal unit test that builds inside of the runtime tests the randomization and the test against the dll with Q#->QIR just verifies that the Q# callables work as expected without trying to validate the randomness of the results (ie: check that a number is returned but don't try to validate uniqueness). This one is the most work but gives us the most comprehensive test coverage.

I'll file an issue for this as pending clean-up work so it doesn't get lost.

Reply:
Filed as #599

Reply:
Some time later I will need to take a more detailed look at the new structure introduced by the earlier Cesar's PR and this PR, to get a broader picture.
For now I'm inclined to agree with you that those who will be using the random number generator, will also need to test their functionality, and they will also likely need a 100% deterministic pseudo-random number generators. So it makes sense to expose (`dllexport`) the discussed APIs.
Thank you for clarifying. Let's leave as is for now.